tf_slim/layers/optimizers_test.py

```diff
@@ -0,0 +1,540 @@
+
+  def testWrongOptimizer(self):
+    optimizers = ["blah", variables.Variable, object(), lambda x: None]
+    for optimizer in optimizers:
+      with ops.Graph().as_default() as g:
+        with self.session(graph=g):
+          _, _, loss, global_step = _setup_model()
+          with self.assertRaises(ValueError):
+            optimizers_lib.optimize_loss(
+                loss, global_step, learning_rate=0.1, optimizer=optimizer)
+
+  def testBadSummaries(self):
+    with ops.Graph().as_default() as g, self.session(graph=g):
+      _, _, loss, global_step = _setup_model()
+      with self.assertRaises(ValueError):
+        optimizers_lib.optimize_loss(
+            loss, global_step, learning_rate=0.1, optimizer="SGD",
+            summaries=["loss", "bad_summary"])
+
+  def testInvalidLoss(self):
+    with ops.Graph().as_default() as g, self.session(graph=g):
+      _, _, _, global_step = _setup_model()
+      with self.assertRaises(ValueError):
+        optimizers_lib.optimize_loss(
+            None, global_step, learning_rate=0.1, optimizer="SGD")
+      with self.assertRaises(ValueError):
+        optimizers_lib.optimize_loss(
+            [[1.0]], global_step, learning_rate=0.1, optimizer="SGD")
+
+  def testInvalidGlobalStep(self):
+    with ops.Graph().as_default() as g, self.session(graph=g):
+      x = array_ops.placeholder(dtypes.float32, [])
+      var = variable_scope.get_variable(
+          "test", [], initializer=init_ops.constant_initializer(10))
+      loss = math_ops.abs(var * x)
+      with self.assertRaises(AttributeError):
+        optimizers_lib.optimize_loss(
+            loss,
+            global_step=constant_op.constant(
+                43, dtype=dtypes.int64),
+            learning_rate=0.1,
+            optimizer="SGD")
+      with self.assertRaises(TypeError):
+        optimizers_lib.optimize_loss(
+            loss,
+            global_step=variable_scope.get_variable(
+                "global_step", [],
+                trainable=False,
+                dtype=dtypes.float64,
+                initializer=init_ops.constant_initializer(
+                    0.0, dtype=dtypes.float64)),
+            learning_rate=0.1,
+            optimizer="SGD")
+      with self.assertRaises(ValueError):
+        optimizers_lib.optimize_loss(
+            loss,
+            global_step=variable_scope.get_variable(
+                "global_step", [1],
+                trainable=False,
+                dtype=dtypes.int64,
+                initializer=init_ops.constant_initializer(
+                    [0], dtype=dtypes.int64)),
+            learning_rate=0.1,
+            optimizer="SGD")
+
+  def testInvalidLearningRate(self):
+    with ops.Graph().as_default() as g, self.session(graph=g):
+      _, _, loss, global_step = _setup_model()
+      with self.assertRaises(ValueError):
+        optimizers_lib.optimize_loss(
+            loss, global_step, learning_rate=-0.1, optimizer="SGD")
+
+  def testGradientNoise(self):
```

Comment:
This and the below test seem like they're going to be very flaky and not super reliable.

If you have the time I'd recommend replacing it w/ a test that runs multiple times & checks the mean & std dev:

Reset variables to a fixed constant value after each run, & run it multiple times then test the mean & std dev of the results to make sure if it matches what you expect.

This isn't super high priority though. Since this test is already not great, I'd be okay w/ just deleting it in this PR and coming back to it later if you have the time. (I think it's better to remove this test than leave it in, because it's not a great signal and is likely to break)